src/local_cache.cc

```diff
@@ -215,51 +213,97 @@ LocalCache::Create(
   return nullptr;  // success
 }
 
-bool
-LocalCache::Exists(const std::string& key)
+TRITONSERVER_Error*
+LocalCache::Lookup(
+    const std::string& key, TRITONCACHE_CacheEntry* entry,
+    TRITONCACHE_Allocator* allocator)
 {
   std::unique_lock lk(cache_mu_);
-  return cache_.find(key) != cache_.end();
-}
 
-std::pair<TRITONSERVER_Error*, CacheEntry>
-LocalCache::Lookup(const std::string& key)
-{
-  std::unique_lock lk(cache_mu_);
   // total_lookup_latency must be protected by mutex
   helpers::ScopedTimer timer(total_lookup_latency_us_);
   num_lookups_++;  // must be protected
 
   auto iter = cache_.find(key);
   if (iter == cache_.end()) {
     num_misses_++;  // must be protected
-    auto err = TRITONSERVER_ErrorNew(
+    return TRITONSERVER_ErrorNew(
         TRITONSERVER_ERROR_NOT_FOUND,
         std::string("key [" + key + "] does not exist").c_str());
-    return {err, {}};
   }
 
   num_hits_++;  // must be protected
   UpdateLRU(iter);
-  return std::make_pair(nullptr, iter->second);  // success
+
+  // Build TRITONCACHE_CacheEntry from cache representation of entry
+  const auto& lentry = iter->second;
+  for (const auto [buffer, attrs] : lentry->buffers_) {
+    // Triton will copy from passed attrs if needed, so pass as-is for now
+    size_t byte_size = 0;
+    RETURN_IF_ERROR(
+        TRITONSERVER_BufferAttributesByteSize(attrs.get(), &byte_size));
+    if (!buffer || !attrs || !byte_size) {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INTERNAL,
+          "buffer or attrs was null, or size was zero");
+    }
+
+    // Allocator callback will be used to copy all entry buffers in Triton
+    // before this function returns to avoid pre-mature eviction
+    RETURN_IF_ERROR(
+        TRITONCACHE_CacheEntryAddBuffer(entry, buffer, attrs.get()));
+  }
+
+  // Copy entry buffers directly into allocator-provided buffers
+  return TRITONCACHE_Copy(allocator, entry);
```

Comment:
Just to make sure, I'm understanding this correctly, the reason that we have to call `TRITONCACHE_Copy` in the lookup function is that it needs to be protected by the mutex?

Reply:
Exactly, to protect the cache-owned buffers until they get copied out